text/3634-scoped-impl-trait-for-type.md

```diff
@@ -0,0 +1,3507 @@
+
+Consider *ImplEnvironment* clauses in bounds like here:
+
+```rust
+use another_crate::{Trait, Type1, Type2};
+
+pub fn function<T as Trait in self>() {}
+
+pub use impl Trait for Type1 {}
+pub use impl Trait for Type2 {}
+```
+
+With this construct, `function` could privately rely on implementation details of `Trait` on `Type1` and `Type2` without defining a new sealed wrapper trait. It also becomes possible to easily define multiple sealed sets of implementations this way, by defining modules that export them.
+
+Overall this would act as a more-flexible but also more-explicit counterpart to sealed traits.
+
+Iff the caller is allowed to use this function without restating the binding, then removing the scope would be a breaking change (as it is already with bindings captured on type parameters in public signatures, so that would be consistent for this syntactical shape).
+
+> That convenience (automatically using the correct implementations even if not in scope) also really should exist only iff there already is robust, near-effortless tooling for importing existing scoped implementations where missing. Otherwise this feature here *would* get (ab)used for convenience, which would almost certainly lead to painful overly sealed APIs.
+
+Binding an implementation in a call as `function::<T as Trait in a>()` while it is constrained as `fn function<T as Trait in b>() { … }` MUST fail for distinct modules `a` and `b` even if the implementations are identical, as otherwise this would leak the implementation identity into the set of breaking changes.
+
+## Glue crate suggestions
+[glue-crate-suggestions]: #glue-crate-suggestions
+
+If crates move some of their overlay features into glue crates, as explained in [unblock-ecosystem-evolution], it would be nice if they could suggest them if both they and e.g. Serde were `cargo add`ed as direct dependencies of a crate currently being worked on.
+
+An example of what this could look like:
+
+```toml
+[package]
+name = "my-crate"
+version = "0.1.2"
+edition = "2021"
+
+[dependencies]
+# none
+
+[suggest-with.serde."1"]
+my-crate_serde_glue = "0.1.0"
+
+[suggest-with.bevy_reflect."0.11"]
+my-crate_bevy_reflect_glue = "0.1.2"
+
+[suggest-with.bevy_reflect."0.12"]
+my-crate_bevy_reflect_glue = "0.2.1"
+```
+
+(This sketch doesn't take additional registries into account.)
+
+Ideally, crates.io should only accept existing crates here (but with non-existing version numbers) and Cargo should by default validate compatibility where possible during `cargo publish`.
+
+## Reusable limited-access APIs
+
+Given a newtype of an unsized type, like
+
+```rust
+#[repr(transparent)]
+pub struct MyStr(str);
+```
+
+for example, there is currently no safe-Rust way to convert between `&str` and `&MyStr` or `Box<MyStr>` and `Box<str>`, even though *in the current module which can see the field* this is guaranteed to be a sound operation.
```

Comment:
We need something like this for the delegation proposal as well
https://github.com/rust-lang/rfcs/blob/923b8e65639de80a93ef2b9770ba1c976c24cf5d/text/0000-fn-delegation.md#newtype-support-1
https://github.com/rust-lang/rfcs/blob/923b8e65639de80a93ef2b9770ba1c976c24cf5d/text/0000-fn-delegation.md#newtype-support

Interfaces for going from
- `&Newtype<T>` to `&T`
- `&mut Newtype<T>` to `&mut T`
- `Newtype<T>` to `T`
- `T` to `Newtype<T>`

for `struct Newtype<T>(pub(vis) T)`, *but* not usable outside of `vis` (the wrapped field's visibility area).

For structures that have a single field we can just hardcode such logic into the compiler, but it would be nice to be a bit more extensible.